t/lib/CorpusTest.pm

```diff
@@ -148,109 +129,144 @@ sub _validity_tests {
-
-            if (!$deprecated and $has_canonical_bson and $has_canonical_json) {
-                local $ENV{BSON_EXTJSON} = 1;
-
-                # localized variable
-                my $canonical_json = $canonical_json;
-
-                # fixes for digit tests
-                $canonical_json =~ s{("\$numberDouble"):"-1.0"}{$1:"-1"}g;
-                $canonical_json =~ s{("\$numberDouble"):"1.0"}{$1:"1"}g;
-                $canonical_json =~ s{("\$numberDouble"):"-0.0"}{$1:"0"}g;
-                $canonical_json =~ s{("\$numberDouble"):"0.0"}{$1:"0"}g;
-                $canonical_json =~ s{-1\.23456789012345677E\+18}{-1.23456789012346e+18}g;
-                $canonical_json =~ s{1\.23456789012345677E\+18}{1.23456789012346e+18}g;
-
-                _bson_to_extjson(
-                    $codec,
-                    $canonical_bson,
-                    $canonical_json,
-                    'cB -> cEJ',
-                    0,
-                );
+            my $spec = {
+                canonical_bson => $canonical_bson,
+                converted_bson => $converted_bson,
+                degenerate_bson => $degenerate_bson,
+                canonical_json => $canonical_json,
+                converted_json => $converted_json,
+                degenerate_json => $degenerate_json,
+                relaxed_json => $relaxed_json,
+                lossy => $lossy,
+            };
+
+            if (!$deprecated) {
+                _validity_tests_non_deprecated($codec, $spec);
             }
 
-            if (!$deprecated and $has_canonical_bson and $has_relaxed_json) {
-                my $relaxed_json = $relaxed_json;
-                $relaxed_json =~ s{:-1234567890123456768\}}{:-1.23456789012346e+18\}}g;
-                $relaxed_json =~ s{:1234567890123456768\}}{:1.23456789012346e+18\}}g;
-                _bson_to_extjson(
-                    $codec,
-                    $canonical_bson,
-                    $relaxed_json,
-                    'cB -> rEJ',
-                    1,
-                );
-            }
+            _validity_tests_all($codec, $spec, $deprecated);
+        };
+    }
 
-            if ($has_canonical_json and $has_canonical_bson) {
-                if (!$lossy) {
-                    _extjson_to_bson(
-                        $codec,
-                        $canonical_json,
-                        ($deprecated && $has_converted_bson)
-                            ? $converted_bson
-                            : $canonical_bson,
-                        'cEJ -> cB',
-                    );
-                }
-            }
+    return;
+}
 
-            if (!$deprecated and $has_degenerate_bson and $has_canonical_json) {
-                _bson_to_extjson(
-                    $codec,
-                    $degenerate_bson,
-                    $canonical_json,
-                    'dB -> cEJ',
-                );
-            }
+# deprecated and non-deprecated
+sub _validity_tests_all {
```

Comment:
I still find these very hard to reconcile against the bson-corpus spec.  Could you please try reorganizing to more directly follow the structure in the spec?  (e.g. "For cB input...", "For cEJ input..."?